drivers/Watchdog.h

```diff
@@ -0,0 +1,137 @@
+/** \addtogroup hal */
+/** @{*/
+/* mbed Microcontroller Library
+ * Copyright (c) 2017 ARM Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef MBED_WATCHDOG_H
+#define MBED_WATCHDOG_H
+
+#ifdef DEVICE_WATCHDOG
+
+#include "watchdog_api.h"
+
+#include <cstdio>
+
+
+namespace mbed {
+/** \addtogroup drivers */
+/** @{*/
+/** A system timer that will reset the system in the case of system failures or
+ *  malfunctions.
+ *
+ * Example:
+ * @code
+ *
+ * Watchdog watchdog = Watchdog();
```

Comment:
If multiple watchdog object are created they will all be kicking the same watchdog right? If so you won't be able to protect multiple threads with a watchdog, as several threads may be deadlocked but if at least one is still kicking the the watchdog will not cause a reset.

Reply:
There is a plan for a second version of the Watchdog API that supports multiplexing watchdogs, and I guess some way of letting threads associated themselves with the watchdog and require kicks from each of them to not trigger a reset, but that was out of the scope of this change.

Reply:
Lets reduce the scope of this to only HAL and let Senthil review how to work on the user facing API